Add Parakeet

nithinraok opened the pull request:
# What does this PR do?

This PR adds support for FastConformer encoder and ParakeetCTC models within transformers, enabling the use of NVIDIA's Parakeet and Canary speech-to-text models.

### FastConformer
- **Architecture**: Linearly scalable Conformer encoder for audio processing
- **Features**: Relative positional encoding, efficient attention, convolutional modules
- **Usage**: Can be used as encoder foundation for various speech tasks

### ParakeetCTC  
- **Architecture**: FastConformer encoder + CTC decoder for speech recognition
- **Features**: Complete ASR pipeline with CTC decoding
- **Performance**: State-of-the-art accuracy with computational efficiency


Parakeet-TDT support is coming soon.

nithinraok committed:
first commit

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
update to handle masking for bs>1

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
Add tests and docs

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok commented:
Requesting review from: @eustlb

nithinraok committed:
update model ids

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
update docs and improve style

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
update librosa location

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
import guard torch too

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
ruff code checks fix

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
ruff format check

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

eustlb reviewed (commented):
Thanks a lot for the work @nithinraok, super excited to support parakeet models natively in Transformers 🤗

Let's do a first iteration:
1. we want to have at least one of the standalone usable architectures like `FastConformerForCTC` for example, and idealy have them all so also `FastConformerForTDT` and `FastConformerForRNNT`. 
2. let's rely on [modular](https://huggingface.co/docs/transformers/en/modular_transformers) to ease future maintenance 🤗 I already see that we can inheritate from `LlamaAttention` and `Wav2Vec2ConformerRelPositionalEmbedding` (I would also suggest checking the confermer modeling code for wave2vec2conformer but it is a bit outdated so to be taken with a pinch of salt)

nithinraok commented on src/transformers/models/fastconformer/config.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "model_type": "fastconformer",
+  "vocab_size": 1024,
+  "d_model": 1024,
+  "encoder_layers": 24,
+  "encoder_attention_heads": 8,
+  "encoder_ffn_dim": 4096,
+  "encoder_layerdrop": 0.1,
+  "activation_function": "silu",
+  "dropout": 0.1,
+  "attention_dropout": 0.1,
+  "activation_dropout": 0.1,
+  "initializer_range": 0.02,
+  "conv_kernel_size": 9,
+  "subsampling_factor": 8,
+  "subsampling_conv_channels": 256,
+  "use_bias": false,
+  "num_mel_bins": 128,
+  "pad_token_id": 0,
+  "bos_token_id": 1,
+  "eos_token_id": 2,
+  "architectures": [
+    "FastConformerModel"
+  ]
+} 
```
Makes sense.!

nithinraok committed:
updated to parakeet names

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok commented on tests/models/fastconformer/test_modeling_fastconformer.py:
```diff
@@ -0,0 +1,423 @@
+        self.assertIsNotNone(outputs.last_hidden_state)
+        self.assertEqual(outputs.last_hidden_state.shape[0], self.model_tester.batch_size)
+
+    def test_feature_extractor_integration(self):
```
This is for getting input from feature extractor and passing through model, should it still need to be moved there?

nithinraok commented on src/transformers/models/fastconformer/configuration_fastconformer.py:
```diff
@@ -0,0 +1,143 @@
+            eos_token_id=eos_token_id,
+            **kwargs,
+        )
+
+
```
Added ParakeetCTCConfig to make naming similar to models

nithinraok committed:
update script

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
Add tokenizer decoding

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
Remove other model dependency

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
clean tests

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
fix tests

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
linting

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
fix ruff lint warnings

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

eustlb reviewed (commented):
Thanks for iterating @nithinraok, the PR is in a good shape!

nithinraok committed:
move to seperate folders

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
add parakeet ctc model code

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
simplify encoder structure

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
update documentation

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
add parakeet to toctree

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
fix tests

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
add parakeet doc

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
Address comments

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
Update featurizer to compute lens directly

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
fix ruff tests

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok commented on docs/source/en/model_doc/fastconformer.md:
```diff
@@ -0,0 +1,248 @@
+# Extract mel-spectrogram features
+features = feature_extractor(
+    raw_audio, 
+    audio_lengths=audio_lengths, 
```
Updated featurizer to handle lengths

lizelive commented:
```
D:\source\Dandori\lab\convert_nemo_fastconformer_to_hf.py:122: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tar.extractall(extract_dir)
INFO:__main__:All extracted files: ['248b67f0680446e28a2f445da319615a_vocab.txt', '2e2b4c0cc6bd44d9a379f769249666ae_tokenizer.model', 'ad13e916727a444097e043d5744eaf63_tokenizer.vocab', 'model_config.yaml', 'model_weights.ckpt']
INFO:__main__:Found vocabulary file: 2e2b4c0cc6bd44d9a379f769249666ae_tokenizer.model
INFO:__main__:Found vocabulary file: ad13e916727a444097e043d5744eaf63_tokenizer.vocab
INFO:__main__:Found config file: model_config.yaml
INFO:__main__:Found model weights: model_weights.ckpt
INFO:__main__:Found model files: ['vocab_file', 'model_config', 'model_weights']
ERROR:__main__:Failed to extract and load NeMo file c:\Users\me\Downloads\parakeet-ctc-0.6b.nemo: 'charmap' codec can't decode byte 0x81 in position 2523: character maps to <undefined>
ERROR:__main__:Conversion failed: 'charmap' codec can't decode byte 0x81 in position 2523: character maps to <undefined>
Traceback (most recent call last):
  File "D:\source\Dandori\lab\convert_nemo_fastconformer_to_hf.py", line 833, in <module>
    main()
  File "D:\source\Dandori\lab\convert_nemo_fastconformer_to_hf.py", line 790, in main
    conversion_info = convert_nemo_to_hf(args.path_to_nemo_model, args.output_dir)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\source\Dandori\lab\convert_nemo_fastconformer_to_hf.py", line 598, in convert_nemo_to_hf
    config = load_nemo_config(model_files["model_config"])
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\source\Dandori\lab\convert_nemo_fastconformer_to_hf.py", line 235, in load_nemo_config
    config = yaml.safe_load(f)
             ^^^^^^^^^^^^^^^^^
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\__init__.py", line 79, in load
    loader = Loader(stream)
             ^^^^^^^^^^^^^^
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\loader.py", line 34, in __init__
    Reader.__init__(self, stream)
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\reader.py", line 85, in __init__
    self.determine_encoding()
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\reader.py", line 124, in determine_encoding
    self.update_raw()
  File "C:\Users\me\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\yaml\reader.py", line 178, in update_raw
    data = self.stream.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.12_3.12.2800.0_x64__qbz5n2kfra8p0\Lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 2523: character maps to <undefined>
PS D:\source\Dandori\lab> 
```

conversion fails with `https://huggingface.co/nvidia/parakeet-ctc-0.6b/blob/main/parakeet-ctc-0.6b.nemo`

lizelive reviewed (changes requested):
nice pr, just needs encoding parameter set. doesn't work on windows.

nithinraok committed:
fix encoding format

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

nithinraok committed:
fix minor ctc decoding

Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

lizelive reviewed (changes requested):
the ParakeetCTCTokenizer decoders is broken. 

```
        # Handle empty list
        if not token_ids:
            return ""
 ```

eustlb committed:
revert modular_model_converter.py changes

eustlb committed:
revert check_config_attributes.py changes

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
refactor: fastconformer & parakeet_ctc -> parakeet

eustlb committed:
modeling update

eustlb committed:
test update

eustlb committed:
propagate feature extractor updates

eustlb committed:
propagate doc changes

eustlb committed:
propagate doc changes

eustlb committed:
propagate tokenization changes

eustlb committed:
propagate conversion changes

eustlb committed:
remove fastconformer tests

eustlb committed:
remove modular

eustlb committed:
update processor

eustlb committed:
update processor

eustlb committed:
tset update

eustlb committed:
diverse fixes

eustlb committed:
100% macthing greedy batched

ebezzam committed:
Update conversion script.

ebezzam committed:
Refactor docs.

ebezzam committed:
Reafactor auto loading.

ebezzam committed:
Refactor and fix tokenization and processing.

ebezzam committed:
Update integration test.

ebezzam committed:
Modeling fixes:
- ensure correct attention mask shape
- ensure layer drop returns valid output
- correct blank token ID when computing CTC loss

ebezzam committed:
Format and repo consistency.

ebezzam committed:
Update model doc.

ebezzam committed:
Fix feature extraction tests.

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        return hidden_states
+
+
+class ParakeetEncoderConvModule(nn.Module):
```
`FastSpeech2ConformerConvolutionModule` seems to be close, if we can standardize that would be nice?

ArthurZucker replied:
Gemma3n as well

ebezzam replied:
I've standardized `FastSpeech2ConformerConvolutionModule` so it can be used by Parakeet, and checked that it introduced not new errors with `RUN_SLOW=1 pytest tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py`. There are two failing tests from before the change:
```python
FAILED tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py::FastSpeech2ConformerModelTest::test_multi_gpu_data_parallel_forward - RuntimeError: Input tensor at index 1 has invalid shape [3, 3, 8], but expected [3, 14, 8]
FAILED tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py::FastSpeech2ConformerWithHifiGanTest::test_multi_gpu_data_parallel_forward - RuntimeError: Input tensor at index 1 has invalid shape [3, 11, 8], but expected [3, 4, 8]
======== 2 failed, 88 passed, 170 skipped, 2 warnings in 31.24s =========
```

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        return hidden_states.transpose(1, 2)
+
+
+def eager_attention_forward(
+    module: nn.Module,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attention_mask: Optional[torch.Tensor],
+    scaling: float,
+    dropout: float = 0.0,
+    **kwargs: Unpack[TransformersKwargs],
+):
+    attn_weights = torch.matmul(query, key.transpose(2, 3)) * scaling
+    if attention_mask is not None:
+        attn_weights = attn_weights + attention_mask
+
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+    attn_output = torch.matmul(attn_weights, value)
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attn_weights
```
using modular this can be imported

ebezzam replied:
@eustlb had a [modular script](https://github.com/eustlb/transformers/blob/add_fastconformer-changes/src/transformers/models/parakeet/modular_parakeet.py) in his fork? Maybe he had a reason for removing it when pushing changes here?

Anyway I just had to add `self.num_key_value_groups` to `ParakeetEncoderAttention` to be able to use `eager_attention_forward` from `src/transformers/models/llama/modeling_llama.py`

But I agree, there's probably a cleaner way to have `ParakeetEncoderAttention` copied from `LlamaAttention` with modular...

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        if self.config._attn_implementation != "eager":
+            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        if position_embeddings is not None:
+            query_states_with_bias_u = query_states + self.bias_u.view(
+                1, self.config.num_attention_heads, 1, self.head_dim
+            )
+            query_states_with_bias_v = query_states + self.bias_v.view(
+                1, self.config.num_attention_heads, 1, self.head_dim
+            )
+
+            relative_key_states = self.relative_k_proj(position_embeddings)
+            relative_key_states = relative_key_states.view(1, -1, self.config.num_attention_heads, self.head_dim)
+
+            # terms (b) and (d)
+            matrix_bd = query_states_with_bias_v @ relative_key_states.permute(0, 2, 3, 1)
+            matrix_bd = self.rel_shift(matrix_bd)
+            matrix_bd = matrix_bd[..., :seq_length]
+            matrix_bd = matrix_bd * self.scaling
+
+            if attention_mask is not None:
+                matrix_bd = matrix_bd.masked_fill_(attention_mask.logical_not(), -10000.0)
```
is there an attetnion layer without pos emb? We can split into 2 different branches

eustlb replied:
There's no actually, I'll update with `positional_embeddings` being a required arg

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+
+        # to match the original implementation
+        # cf https://github.com/NVIDIA/NeMo/blob/620d2ba07835c230b2e1ee25fe1322504ce01d79/nemo/collections/asr/parts/submodules/multi_head_attention.py#L336-L340
+        if attention_mask is not None:
+            all_masked_rows = torch.all(~attention_mask, dim=-1).transpose(1, 2)
+            all_masked_rows = all_masked_rows.unsqueeze_(-1)
+            attn_output = attn_output.masked_fill(all_masked_rows, 0.0)
```
this should only be applied for eager and sdpa but not for flash as pads are ignored.

ebezzam replied:
should the condition be modified to `if attention_mask is not None and self.config._attn_implementation in ["eager", "sdpa"]`?

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+    def rel_shift(self, attention_scores):
+        """Relative position shift for Shaw et al. style attention."""
+        batch_size, num_heads, query_length, position_length = attention_scores.shape
+        attention_scores = nn.functional.pad(attention_scores, pad=(1, 0))
```
so we pad in the position length on the right, but then we view something that was:
`batch_size, num_heads, query_length, position_length` to `batch_size, num_heads, position_length+, query_length`, and then we ignore the first one (the padding). 
Would be nice to have a small intuitive explanaiton of what is happening here!

ebezzam replied:
@eustlb?

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+            pointwise_conv = nn.Conv2d(self.channels, self.channels, kernel_size=1)
+            layers.extend([depthwise_conv, pointwise_conv, nn.ReLU()])
+
+        self.layers = nn.Sequential(*layers)
```
we don't use sequential let's use modulelist

ArthurZucker replied:
also why not create a layer for dept + point _ relu?

ebezzam replied:
Switched to ModuleList.

Layer for (dept + point + relu) could be nice, but the forward for `ParakeetEncoderSubsamplingConv2D` would need a not-so-intuitive forward logic. Atm attention mask is applied after each Conv2D layer:
```python
for layer in self.layers:
        hidden_states = layer(hidden_states)

        # mask the hidden states
        if isinstance(layer, nn.Conv2d) and attention_mask is not None:
            current_lengths = self._get_output_length(current_lengths, layer)
            current_seq_length = hidden_states.shape[2]
            channel_mask = (
                torch.arange(current_seq_length, device=attention_mask.device) < current_lengths[:, None]
            )
            hidden_states *= channel_mask[:, None, :, None]
```
But if one of the layers is a combined (dept + point + relu), would still need to call each sublayer independently to apply the mask in between

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        )
+
+        if attention_mask is not None:
+            # TODO: @eustlb, which mask utils to do the same?
```
yes `create_causal_mask` `from ...masking_utils import create_causal_mask, create_sliding_window_causal_mask`

ebezzam replied:
Thanks, tried `create_causal_mask` and output shape of `attention_mask` is correct so it's functional. But predicted token IDs have large mismatch with expected ones. Need to figure it out why...

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+            # ctc_loss doesn't support fp16
+            log_probs = nn.functional.log_softmax(logits, dim=-1, dtype=torch.float32).transpose(0, 1)
+
+            with torch.backends.cudnn.flags(enabled=False):
```
if cuddn is not available this will fail no?

ebezzam replied:
Seems to be the approach done for other CTC models? (wav2vec2, wavlm, hubert, etc) https://github.com/huggingface/transformers/blob/8b804311ba9bf921a6099c32edb6aae1ae557b8e/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L1897

ArthurZucker commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+            **kwargs,
+        )
+
+        # greedy decoding
+        sequences = outputs.logits.argmax(dim=-1)
+
+        # mask out padded tokens
+        if attention_mask is not None:
+            attention_mask = self._get_output_attention_mask(attention_mask)
+            sequences[~attention_mask] = self.config.pad_token_id
+
```
I do find this a bit weird but if that's what it takes to enable

ebezzam committed:
Fix (most) tokenizer tests.

ebezzam committed:
Add pipeline example.

ebezzam commented on src/transformers/models/parakeet/convert_nemo_to_hf.py:
```diff
@@ -0,0 +1,950 @@
+        sample_lines = lines[:5]
+        has_tabs = any("\t" in line for line in sample_lines)
+
+        if has_tabs:
+            logger.info("Detected SentencePiece format (token\\tscore)")
+            # SentencePiece vocab format: token \t score
+            for token_id, line in enumerate(lines):
+                parts = line.strip().split("\t")
+                if len(parts) >= 1 and parts[0]:  # Ensure token is not empty
+                    token = parts[0]
+                    vocab_dict[token] = token_id
+        else:
+            logger.info("Detected plain text format (one token per line)")
+            # Plain text format: one token per line
+            for token_id, line in enumerate(lines):
+                token = line.strip()
+                if token:  # Ensure token is not empty
+                    vocab_dict[token] = token_id
+
+        logger.info(f"Successfully converted vocabulary with {len(vocab_dict)} tokens")
+
+        # Ensure <unk> token exists
+        if "<unk>" not in vocab_dict:
+            # Insert <unk> at the beginning and shift other tokens
+            new_vocab_dict = {"<unk>": 0}
+            for token, token_id in vocab_dict.items():
+                new_vocab_dict[token] = token_id + 1
+            vocab_dict = new_vocab_dict
+            logger.info("Added <unk> token at ID 0 and shifted other tokens")
+
+        return vocab_dict
+
+    except Exception as e:
+        logger.error(f"Failed to convert vocab file {vocab_file_path}: {e}")
+        import traceback
+
+        logger.error(f"Traceback: {traceback.format_exc()}")
+        # Return a minimal vocab as fallback
+        logger.warning("Creating minimal fallback vocabulary")
+        return {"<unk>": 0}
```
Do you mean to this this file? https://github.com/huggingface/transformers/blob/main/src/transformers/convert_slow_tokenizer.py

ebezzam commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        positions = torch.arange(length - 1, -length, -1, dtype=torch.float32, device=device).unsqueeze(1)
+        self.create_pe(positions=positions, dtype=dtype)
+
+    def create_pe(self, positions: torch.Tensor, dtype: torch.dtype):
+        """Create positional encoding matrix."""
+        pe = torch.zeros(positions.size(0), self.hidden_size, dtype=dtype, device=positions.device)
+        div_term = torch.exp(
+            torch.arange(0, self.hidden_size, 2, dtype=torch.float32, device=positions.device)
+            * -(math.log(10000.0) / self.hidden_size)
+        )
+        pe[:, 0::2] = torch.sin(positions * div_term)
+        pe[:, 1::2] = torch.cos(positions * div_term)
+
+        # Register as buffer and add batch dimension
+        pe_tensor = pe.unsqueeze(0)  # (1, T, D)
+        self.register_buffer("pe", pe_tensor, persistent=False)
```
Were you thinking something from [modeling_llama](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py) or [modeling_rope_utils](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py)? I couldn't see something exactly the same since Llama (from what I understand) is using rotary PE while Parakeet is using conventional/additive PE

eustlb replied:
It's a bit far from regular ROPE actually and rather close from absolute sinusoidal positional (eg. whisper) embeddings, with the difference here of being relative to the meaning of from a position to another rather than absolute.

ebezzam commented on src/transformers/models/parakeet/modeling_parakeet.py:
```diff
@@ -0,0 +1,602 @@
+        center_pos = self.pe.shape[1] // 2 + 1
+        start_pos = center_pos - seq_length
+        end_pos = center_pos + seq_length - 1
+        position_embeddings = self.pe[:, start_pos:end_pos]
```
`self.extend_pe` may not be cuda compatible due to dynamic memory allocation but slicing on pre-allocated tensor should be fine? I've updated to this.

eustlb replied:
okay for me to have it pre-allocated

ebezzam committed:
Fixes

ebezzam committed:
Use eager_attention_forward from Llama.

ebezzam committed:
Small tweaks.

ebezzam committed:
Replace Sequential with ModuleList

ebezzam committed:
Add check if not all layers copied

ebezzam committed:
Clean tokenizer.

ebezzam committed:
Standardize FastSpeech2ConformerConvolutionModule for Parakeet.

ebezzam commented on src/transformers/models/parakeet/tokenization_parakeet.py:
If you have a pointer/example on how to do this, it would be helpful!

ebezzam committed:
Switch to modular for modeling and processing.

ebezzam committed:
Add processor tests.

ebezzam committed:
Fix modeling tests.

ebezzam committed:
Formating and docstrings.

ebezzam commented on tests/models/parakeet/test_modeling_parakeet.py:
```diff
@@ -0,0 +1,392 @@
+    def test_model_get_set_embeddings(self):
+        pass
+
+    @unittest.skip(reason="Flash not yet supported in Parakeet models")
+    def test_sdpa_can_dispatch_on_flash(self):
+        pass
```
`_supports_flash_attn = False` so not sure why this test gets called?

ebezzam commented:
run-slow: fastspeech2_conformer, parakeet

ebezzam committed:
Merge branch 'main' into add_fastconformer

ebezzam committed:
Add `return_attention_mask` like other feature extractors.

ebezzam committed:
Merge branch 'main' into add_fastconformer

ebezzam committed:
clean up after merging main.

ebezzam committed:
Merge branch 'add_fastconformer' of github.com:nithinraok/transformers into add_fastconformer

ebezzam commented:
run-slow: fastspeech2_conformer, parakeet

ebezzam commented on src/transformers/models/parakeet/modular_parakeet.py:
```diff
@@ -0,0 +1,593 @@
+        if attention_mask is not None:
+            attention_mask = self._get_output_attention_mask(attention_mask, target_length=hidden_states.shape[1])
+
+            # NOTE: @eustlb, which mask utils to do the same?
+            # @ebezzam could use `create_causal_mask` but more complicated
+            # as `and_mask_function` requires callable mask function
+            attention_mask = attention_mask.unsqueeze(1).expand(-1, hidden_states.shape[1], -1)
+            attention_mask = attention_mask & attention_mask.transpose(1, 2)
+            attention_mask = attention_mask.unsqueeze(1)
+            # attention_mask = create_causal_mask(
+            #     self.config,
+            #     hidden_states,
+            #     attention_mask,
+            #     cache_position=torch.arange(attention_mask.shape[1], device=hidden_states.device),
+            #     past_key_values=None,
+            #     position_ids=None,
+            #     # and_mask_function=None,
+            # )
+            # # ensure boolean mask because of `~attention_mask` in later processing
+            # if attention_mask is not None and attention_mask.dtype != torch.bool:
+            #     attention_mask = attention_mask > 0
```
@ArthurZucker `create_causal_mask` seems to be clunkier to use than their custom processing, because of this AND operation on 430. 

Seems like `create_causal_mask` has an `and_mask_function` but it has to be a callable function

eustlb reviewed (commented):
A quick review for myself since I'll be finishing this PR!

ebezzam commented on src/transformers/models/parakeet/configuration_parakeet.py:
```diff
@@ -0,0 +1,251 @@
+        self.hidden_size = hidden_size
+        self.num_hidden_layers = num_hidden_layers
+        self.num_attention_heads = num_attention_heads
+        self.num_key_value_heads = num_attention_heads  # LlamaAttention compatibility
```
Sounds good. In Llama they have `num_key_value_heads=None` in __init__ and add below:

```
# for backward compatibility
if num_key_value_heads is None:
    num_key_value_heads = num_attention_heads
```

eustlb committed:
nits on modeling

eustlb committed:
configuration update

eustlb committed:
nit

eustlb committed:
simplification: use PretrainedTokenizerFast, simplify processor

eustlb committed:
add dtype arg to mel_filter_bank

eustlb committed:
feature extraction: simplify!

eustlb committed:
modeling update

eustlb committed:
change to ParakeetTokenizerFast

eustlb committed:
correct attention mask handling

eustlb committed:
auto update

eustlb committed:
proc update

eustlb committed:
test update

eustlb committed:
feature extraction fixes

eustlb committed:
modeling update

eustlb committed:
conversion script update

eustlb committed:
udpate tests feature integration

eustlb committed:
update tokenization and tests

eustlb committed:
processor tests

eustlb committed:
revert audio_utils

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
config docstring update

eustlb committed:
blank_token -> pad_token

eustlb committed:
modeling udpate

eustlb committed:
doc update

eustlb committed:
fix tests

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
fix test

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
fix tests

ArthurZucker reviewed (approved):
Nice, a few small nits! but 🧼 
About attention, I remember we had "issues". But looking at it now it should work no? 

Great work everyone! 🚀

eustlb committed:
address review comments

eustlb committed:
add comment

eustlb commented on src/transformers/models/parakeet/modular_parakeet.py:
```diff
@@ -0,0 +1,634 @@
+        hidden_states = hidden_states + conv_output
+
+        ff2_output = self.feed_forward2(self.norm_feed_forward2(hidden_states))
+        hidden_states = hidden_states + 0.5 * ff2_output
```
The conformer architecture comes with an fc factor of 0.5. It's architecture-relative and should therefore be hardcodec this way. Adding a comment

eustlb committed:
add comment

eustlb commented on src/transformers/models/parakeet/modular_parakeet.py:
```diff
@@ -0,0 +1,634 @@
+        # will compute matrix_ac - terms (a) and (c) - and add matrix_bd
+        attn_output, attn_weights = attention_interface(
+            self,
+            query_states_with_bias_u,
+            key_states,
+            value_states,
+            matrix_bd,
```
Indeed, supports alibi but not custom attention bias yet.
Explicitly setting `_supports_flash_attn=False` and adding a comment for when custom bias is in flash attention will be available.

eustlb committed:
explicitly not support flash

eustlb committed:
atttention straightforward masking

eustlb commented on src/transformers/models/parakeet/modular_parakeet.py:
```diff
@@ -0,0 +1,634 @@
+        if attention_mask is not None:
+            all_masked_rows = torch.all(~attention_mask, dim=-1).transpose(1, 2)
+            all_masked_rows = all_masked_rows.unsqueeze_(-1)
+            attn_output = attn_output.masked_fill(all_masked_rows, 0.0)
```
was unnecessary with proper masking of the attention mask with `float(-inf)` values. This diverges a bit from original codebase but should be equivalent. We'll very WER perfs to confirm!

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
fix

eustlb committed:
tokenizer update: skipping blank tokens by default

eustlb committed:
doc update

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
fix max_positions_embeddings handling

eustlb committed:
nits

eustlb committed:
change atol faeture extraction integration tests

eustlb committed:
doc update + fix loss

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
doc update

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
nit

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
update integration test for A10

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
repo id name

eustlb committed:
Merge branch 'main' into add_fastconformer

eustlb committed:
nit

eustlb committed:
Merge branch 'main' into add_fastconformer
